libraries/GCS_MAVLink/GCS_Rally.cpp

```diff
@@ -101,3 +102,73 @@ void GCS_MAVLINK::handle_common_rally_message(mavlink_message_t *msg)
     }
 }
 
+MAV_MISSION_RESULT MissionItemProtocol_Rally::get_item(const GCS_MAVLINK &_link,
+                                                       const mavlink_message_t &msg,
+                                                       const mavlink_mission_request_int_t &packet,
+                                                       mavlink_mission_item_int_t &ret_packet)
+{
+    RallyLocation rallypoint;
+
+    if (!rally.get_rally_point_with_index(packet.seq, rallypoint)) {
+        return MAV_MISSION_INVALID_SEQUENCE;
+    }
+
+    ret_packet.command = MAV_CMD_NAV_RALLY_POINT;
+    ret_packet.x = rallypoint.lat;
+    ret_packet.y = rallypoint.lng;
+    ret_packet.z = rallypoint.alt;
+
+    return MAV_MISSION_ACCEPTED;
+}
+
+void MissionItemProtocol_Rally::set_count(const mavlink_mission_count_t &packet)
+{
+    expected_item_count = packet.count;
+}
+void MissionItemProtocol_Rally::timeout()
+{
+    link->send_text(MAV_SEVERITY_WARNING, "Rally upload timeout");
+    expected_item_count = -1;
+}
+
+void MissionItemProtocol_Rally::convert_MISSION_ITEM_INT_to_RallyLocation(const mavlink_mission_item_int_t &cmd, RallyLocation &ret) const
+{
+    ret.lat = cmd.x;
+    ret.lng = cmd.y;
+    ret.alt = constrain_int32(cmd.z, INT16_MIN, INT16_MAX);
+}
+
+MAV_MISSION_RESULT MissionItemProtocol_Rally::replace_item(const mavlink_mission_item_int_t &cmd)
+{
+    RallyLocation rallyloc;
+    convert_MISSION_ITEM_INT_to_RallyLocation(cmd, rallyloc);
+    if (!rally.set_rally_point_with_index(cmd.seq, rallyloc)) {
+        return MAV_MISSION_ERROR;
+    }
+    return MAV_MISSION_ACCEPTED;
+}
+
+MAV_MISSION_RESULT MissionItemProtocol_Rally::append_item(const mavlink_mission_item_int_t &cmd)
+{
+    RallyLocation rallyloc;
+    convert_MISSION_ITEM_INT_to_RallyLocation(cmd, rallyloc);
+    const uint8_t current_total = rally.get_rally_total();
+    if (!rally.set_rally_point_with_index(current_total, rallyloc)) {
```

Comment:
This looks wrong. `AP_Rally::set_rally_point_with_index()` will reject the set if the index is `>=` the total, and you are fixing the total afterwards. I suspect that do to desires to do threading in the future, and to ease other libraries like scripting creating a function in `AP_Rally` that manages the index + adding rather then having the protocol handler know the inner dirty secrets would make more sense.

Reply:
> This looks wrong. AP_Rally::set_rally_point_with_index() will reject the set if the index is >=

No, it really doesn't.  Distinguish between "rally_total" and "rally_max"....